source/common/upstream/upstream_impl.h

```diff
@@ -634,6 +638,7 @@ class ClusterInfoImpl : public ClusterInfo, protected Logger::Loggable<Logger::I
   Stats::ScopePtr stats_scope_;
   mutable ClusterStats stats_;
   Stats::IsolatedStoreImpl load_report_stats_store_;
+  const absl::optional<ClusterRequestResponseSizeStats> request_response_size_stats_;
```

Comment:
256 extra bytes for an absl::optional seems egregious. Is it reserving the entire space for the object, rather than a pointer? 

My suggestion is to (a) use a unique_ptr rather than absl::optional here -- you can still return the optional in the accessor above, and (b) combine this and ClusterTimeoutBudgetStats below into a single; possibly nested structure.

That way the default case will take minimal overhead of just one pointer for all aggregated optional stats.

Reply:
> Is it reserving the entire space for the object, rather than a pointer?

Qouting from an [article](https://abseil.io/tips/123), `empty absl::optional uses as much memory as a populated one`. It says that that memory usage of `absl::optional<Bar>` is always `sizeof(Bar) + sizeof(bool)` irrespective of whether it is null or not. However, with `std::unique_ptr<Bar>`, memory usage is ` sizeof(Bar*) when null` and `sizeof(Bar*) + sizeof(Bar) otherwise`

Reply:
> you can still return the optional in the accessor above

Is this to make code changes minimal?

Reply:
> one pointer for all aggregated optional stats

```
struct OptionalStats{
    std::unique_ptr<ClusterRequestResponseSizeStats> p1;
    std::unique_ptr<ClusterTimeoutBudgetStats> p2;
    ...
};

std::unique_ptr<OptionalStats> p3;
```
Something like above? which would occupy 8 bytes for `unique_ptr<OptionalStats >` in general case.

I guess this would also mean we need to have a nested definition for the config of Optional stats as well, something like `track_optional_stats: {timeout_budget: true, request_response_sizes: true}`.

Reply:
RE "still return the optional" -- it's not to minimize code changes; that makes sense at the interface. But it should be an optional reference, not a reference to the optional object.

So, for. your sample code, `std::unique_ptr<OptionalStats> p3;` looks perfect.

I think inside your OptionalStats struct you could use optionals or unique_ptr. I'm OK either way. Main disadvantage of unique_ptr inside is the double-indirection, but it's probably fine.

Reply:
nit: maybe name it `OptionalClusterStats`

Reply:
Thank you..I think using unique_ptr inside strict would be better as the use cases of each optional stat are quite different. And if we use optional inside struct then it would add up memory for all stats even if we need to track a subset of stats.